In [83]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.metrics import confusion_matrix, accuracy_score

In [84]:
from google.colab import drive
drive.mount ('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [85]:
# Importing File
raw_data = pd.read_csv('/content/drive/MyDrive/Project_placement/RUL/CMaps/test_FD004.txt',sep=' ', header = None)
raw_data

0    1        2       3      4       5       6        7        8   \
0        1    1  20.0072  0.7000  100.0  491.19  606.67  1481.04  1227.81   
1        1    2  24.9984  0.6200   60.0  462.54  536.22  1256.17  1031.48   
2        1    3  42.0000  0.8420  100.0  445.00  549.23  1340.13  1105.88   
3        1    4  42.0035  0.8402  100.0  445.00  549.19  1339.70  1107.26   
4        1    5  35.0079  0.8400  100.0  449.44  555.10  1353.04  1117.80   
...    ...  ...      ...     ...    ...     ...     ...      ...      ...   
41209  248  277  41.9991  0.8401  100.0  445.00  550.30  1364.40  1129.17   
41210  248  278  20.0026  0.7005  100.0  491.19  608.00  1494.75  1260.88   
41211  248  279  34.9988  0.8413  100.0  449.44  555.92  1370.65  1130.97   
41212  248  280  20.0027  0.7000  100.0  491.19  608.19  1489.11  1256.25   
41213  248  281  35.0075  0.8402  100.0  449.44  556.40  1378.58  1140.70   

         9   ...       18       19    20   21    22      23     24       25  \
0      9.35  ...  8048.98   9.2229  0.02  362  2324  100.00  24.31  14.7007   
1      7.05  ...  7863.46  10.8632  0.02  306  1915   84.93  14.36   8.5748   
2      3.91  ...  8071.13   9.3960  0.02  328  2212  100.00  10.39   6.4365   
3      3.91  ...  8078.89   9.3594  0.02  328  2212  100.00  10.56   6.2367   
4      5.48  ...  8057.83   9.3030  0.02  333  2223  100.00  14.85   8.9326   
...     ...  ...      ...      ...   ...  ...   ...     ...    ...      ...   
41209  3.91  ...  8112.61   9.4427  0.02  331  2212  100.00  10.53   6.2620   
41210  9.35  ...  8086.83   9.2772  0.02  366  2324  100.00  24.33  14.6486   
41211  5.48  ...  8100.84   9.3982  0.02  336  2223  100.00  14.69   8.8389   
41212  9.35  ...  8085.24   9.2727  0.03  366  2324  100.00  24.44  14.6887   
41213  5.48  ...  8098.17   9.3964  0.02  335  2223  100.00  14.72   8.8502   

       26  27  
0     NaN NaN  
1     NaN NaN  
2     NaN NaN  
3     NaN NaN  
4     NaN NaN  
...    ..  ..  
41209 NaN NaN  
41210 NaN NaN  
41211 NaN NaN  
41212 NaN NaN  
41213 NaN NaN  

[41214 rows x 28 columns]

In [86]:
raw_data = raw_data.drop([26,27], axis='columns')
raw_data.columns = ['ID', 'Cycle', 'OpSet1', 'OpSet2', 'OpSet3', 'SensorMeasure1', 'SensorMeasure2', 'SensorMeasure3', 'SensorMeasure4', 'SensorMeasure5', 'SensorMeasure6', 'SensorMeasure7', 'SensorMeasure8', 'SensorMeasure9', 'SensorMeasure10', 'SensorMeasure11', 'SensorMeasure12', 'SensorMeasure13', 'SensorMeasure14', 'SensorMeasure15', 'SensorMeasure16', 'SensorMeasure17', 'SensorMeasure18', 'SensorMeasure19', 'SensorMeasure20', 'SensorMeasure21']

In [87]:
raw_data.head(50)

ID  Cycle   OpSet1  OpSet2  OpSet3  SensorMeasure1  SensorMeasure2  \
0    1      1  20.0072  0.7000   100.0          491.19          606.67   
1    1      2  24.9984  0.6200    60.0          462.54          536.22   
2    1      3  42.0000  0.8420   100.0          445.00          549.23   
3    1      4  42.0035  0.8402   100.0          445.00          549.19   
4    1      5  35.0079  0.8400   100.0          449.44          555.10   
5    1      6  25.0010  0.6203    60.0          462.54          536.40   
6    1      7  19.9996  0.7015   100.0          491.19          607.29   
7    1      8  35.0015  0.8414   100.0          449.44          554.81   
8    1      9  19.9980  0.7000   100.0          491.19          607.35   
9    1     10  42.0033  0.8416   100.0          445.00          549.30   
10   1     11  19.9993  0.7000   100.0          491.19          606.40   
11   1     12   9.9999  0.2519   100.0          489.05          604.29   
12   1     13  19.9993  0.7012   100.0          491.19          606.82   
13   1     14   9.9990  0.2516   100.0          489.05          604.16   
14   1     15  35.0060  0.8400   100.0          449.44          555.01   
15   1     16  19.9985  0.7000   100.0          491.19          606.96   
16   1     17  42.0040  0.8400   100.0          445.00          549.30   
17   1     18  41.9986  0.8400   100.0          445.00          549.25   
18   1     19  42.0001  0.8419   100.0          445.00          548.72   
19   1     20  10.0077  0.2518   100.0          489.05          604.28   
20   1     21   0.0025  0.0000   100.0          518.67          641.71   
21   1     22  42.0046  0.8400   100.0          445.00          549.17   
22   1     23   9.9999  0.2500   100.0          489.05          604.11   
23   1     24   0.0003  0.0000   100.0          518.67          641.75   
24   1     25  35.0036  0.8406   100.0          449.44          554.86   
25   1     26  24.9989  0.6200    60.0          462.54          536.31   
26   1     27  19.9990  0.7000   100.0          491.19          606.51   
27   1     28  42.0041  0.8412   100.0          445.00          549.28   
28   1     29   0.0017  0.0007   100.0          518.67          641.78   
29   1     30  35.0029  0.8400   100.0          449.44          555.60   
30   1     31  20.0003  0.7005   100.0          491.19          607.60   
31   1     32  10.0070  0.2500   100.0          489.05          604.29   
32   1     33  42.0075  0.8400   100.0          445.00          549.46   
33   1     34  25.0063  0.6203    60.0          462.54          536.17   
34   1     35  20.0003  0.7000   100.0          491.19          606.91   
35   1     36  42.0056  0.8400   100.0          445.00          548.82   
36   1     37  42.0062  0.8402   100.0          445.00          549.36   
37   1     38  25.0013  0.6218    60.0          462.54          536.54   
38   1     39  35.0051  0.8417   100.0          449.44          554.85   
39   1     40  35.0018  0.8400   100.0          449.44          555.43   
40   1     41  24.9986  0.6207    60.0          462.54          535.76   
41   1     42  25.0062  0.6200    60.0          462.54          536.39   
42   1     43   0.0024  0.0000   100.0          518.67          642.20   
43   1     44  42.0005  0.8400   100.0          445.00          548.90   
44   1     45  25.0020  0.6200    60.0          462.54          536.13   
45   1     46  19.9981  0.7000   100.0          491.19          606.62   
46   1     47  42.0006  0.8419   100.0          445.00          548.41   
47   1     48  20.0050  0.7000   100.0          491.19          606.98   
48   1     49  41.9998  0.8400   100.0          445.00          549.01   
49   1     50   0.0023  0.0005   100.0          518.67          642.02   

    SensorMeasure3  SensorMeasure4  SensorMeasure5  ...  SensorMeasure12  \
0          1481.04         1227.81            9.35  ...           313.03   
1          1256.17         1031.48            7.05  ...           163.61   
2

In [88]:
given_no_of_cycles=[]
for i in range (1,len(raw_data['ID'].unique())+1):
    given_no_of_cycles.append(len(raw_data[raw_data['ID']==i]))
    print(len(raw_data[raw_data['ID']==i]))

given_no_of_cycles=np.array(given_no_of_cycles).reshape([cycle_ran_after.shape[0],1])

230
153
141
208
51
146
54
248
284
23
82
270
227
102
168
240
334
247
24
115
202
158
162
174
486
90
144
20
216
229
134
260
107
361
278
159
345
125
214
266
414
200
120
406
78
37
224
214
273
206
137
58
67
81
103
180
121
185
37
135
396
240
158
251
186
97
150
241
162
190
292
118
149
223
203
48
108
106
184
112
60
146
274
88
174
254
167
140
285
118
144
116
200
105
110
123
51
82
237
196
166
443
145
65
351
148
121
145
66
348
342
88
132
218
82
161
97
178
228
170
194
117
236
230
21
220
98
139
315
107
193
115
74
266
435
56
291
239
88
160
19
69
327
34
102
84
116
126
320
202
285
195
130
122
161
28
138
171
160
122
257
99
122
29
178
63
99
74
34
48
355
132
151
135
116
185
146
279
262
367
199
98
125
154
203
47
121
157
107
200
71
203
263
214
186
70
196
115
177
115
139
222
87
19
95
291
179
180
171
135
206
261
288
102
72
264
141
63
110
65
176
178
305
152
39
59
62
126
20
157
230
370
111
179
362
176
221
177
28
204
42
112
199
184
63
29
94
281


ValueError: ignored

We can see that data for 31 cycles are available for engine 1 in test set. Data for 49 cycles are available for engine 2, and so on. Using the available data our goal is to predict RUL for each engine.

In [89]:
cycle_ran_after = pd.read_csv('/content/drive/MyDrive/Project_placement/RUL/CMaps/RUL_FD004.txt', sep=' ', header = None)
cycle_ran_after=cycle_ran_after.drop([1], axis='columns')
cycle_ran_after = np.array(cycle_ran_after)
cycle_ran_after

array([[ 22],
       [ 39],
       [107],
       [ 75],
       [149],
       [ 78],
       [ 94],
       [ 14],
       [ 99],
       [162],
       [143],
       [  7],
       [ 71],
       [105],
       [ 12],
       [160],
       [162],
       [104],
       [194],
       [ 82],
       [ 91],
       [ 11],
       [ 26],
       [142],
       [ 39],
       [ 92],
       [ 76],
       [124],
       [ 64],
       [118],
       [  6],
       [ 22],
       [147],
       [126],
       [ 36],
       [ 73],
       [ 89],
       [ 11],
       [151],
       [ 10],
       [ 97],
       [ 30],
       [ 42],
       [ 60],
       [ 85],
       [134],
       [ 34],
       [ 45],
       [ 24],
       [ 86],
       [119],
       [151],
       [142],
       [176],
       [157],
       [ 67],
       [ 97],
       [  8],
       [154],
       [139],
       [ 51],
       [ 33],
       [184],
       [ 46],
       [ 12],
       [133],
       [ 46],
       [ 46],
       [ 12],
       [ 33],
       [ 15],
      

An RUL value of 112 for engine 1 in test set means that it failed 112 cycles after 31st cycle. So engine 1 ran for (31+112) cycles before it failed. Similarly engine two failed 98 cycles after 49th cycle. So engine 2 ran for (49 + 98) cycles in total before it failed. Similar calculation can be done for other engines in test set.

#Making EOL column

In [90]:
EOL=[]
i=0

for i in raw_data['ID']:
        EOL.append(given_no_of_cycles[i-1]+cycle_ran_after[i-1] )
EOL = np.array(EOL).reshape([len(EOL),1])

In [91]:
raw_data['EOL']=EOL
raw_data

ID  Cycle   OpSet1  OpSet2  OpSet3  SensorMeasure1  SensorMeasure2  \
0        1      1  20.0072  0.7000   100.0          491.19          606.67   
1        1      2  24.9984  0.6200    60.0          462.54          536.22   
2        1      3  42.0000  0.8420   100.0          445.00          549.23   
3        1      4  42.0035  0.8402   100.0          445.00          549.19   
4        1      5  35.0079  0.8400   100.0          449.44          555.10   
...    ...    ...      ...     ...     ...             ...             ...   
41209  248    277  41.9991  0.8401   100.0          445.00          550.30   
41210  248    278  20.0026  0.7005   100.0          491.19          608.00   
41211  248    279  34.9988  0.8413   100.0          449.44          555.92   
41212  248    280  20.0027  0.7000   100.0          491.19          608.19   
41213  248    281  35.0075  0.8402   100.0          449.44          556.40   

       SensorMeasure3  SensorMeasure4  SensorMeasure5  ...  SensorMeasure13  \
0             1481.04         1227.81            9.35  ...          2387.78   
1             1256.17         1031.48            7.05  ...          2028.09   
2             1340.13         1105.88            3.91  ...          2387.95   
3             1339.70         1107.26            3.91  ...          2387.90   
4             1353.04         1117.80            5.48  ...          2387.87   
...               ...             ...             ...  ...              ...   
41209         1364.40         1129.17            3.91  ...          2388.50   
41210         1494.75         1260.88            9.35  ...          2388.33   
41211         1370.65         1130.97            5.48  ...          2388.64   
41212         1489.11         1256.25            9.35  ...          2388.37   
41213         1378.58         1140.70            5.48  ...          2388.59   

       SensorMeasure14  SensorMeasure15  SensorMeasure16  SensorMeasure17  \
0              8048.98           9.2229             0.02              362   
1              7863.46          10.8632             0.02              306   
2              8071.13           9.3960             0.02              328   
3              8078.89           9.3594             0.02              328   
4              8057.83           9.3030             0.02              333   
...                ...              ...              ...              ...   
41209          8112.61           9.4427             0.02              331   
41210          8086.83           9.2772             0.02              366   
41211          8100.84           9.3982             0.02              336   
41212          8085.24           9.2727             0.03              366   
41213          8098.17           9.3964             0.02              335   

       SensorMeasure18  SensorMeasure19  SensorMeasure20  SensorMeasure21  EOL  
0                 2324           100.00            24.31          14.7007  252  
1                 1915            84.93            14.36           8.5748  252  
2                 2212           100.00            10.39           6.4365  252  
3                 2212           100.00            10.56           6.2367  252  
4                 2223           100.00            14.85           8.9326  252  
...                ...              ...              ...              ...  ...  
41209             2212           100.00            10.53           6.2620  307  
41210             2324           100.00            24.33          14.6486  307  
41211             2223           100.00            14.69           8.8389  307  
41212             2324           100.00            24.44          14.6887  307  
41213             2223           100.00            14.72           8.8502  307  

[41214 rows x 27 columns]

In [92]:
raw_data["LR"] = raw_data["Cycle"]/raw_data["EOL"]
# Drop ID and EOL
raw_data=raw_data.drop(columns=['ID','EOL'])
raw_data

Cycle   OpSet1  OpSet2  OpSet3  SensorMeasure1  SensorMeasure2  \
0          1  20.0072  0.7000   100.0          491.19          606.67   
1          2  24.9984  0.6200    60.0          462.54          536.22   
2          3  42.0000  0.8420   100.0          445.00          549.23   
3          4  42.0035  0.8402   100.0          445.00          549.19   
4          5  35.0079  0.8400   100.0          449.44          555.10   
...      ...      ...     ...     ...             ...             ...   
41209    277  41.9991  0.8401   100.0          445.00          550.30   
41210    278  20.0026  0.7005   100.0          491.19          608.00   
41211    279  34.9988  0.8413   100.0          449.44          555.92   
41212    280  20.0027  0.7000   100.0          491.19          608.19   
41213    281  35.0075  0.8402   100.0          449.44          556.40   

       SensorMeasure3  SensorMeasure4  SensorMeasure5  SensorMeasure6  ...  \
0             1481.04         1227.81            9.35           13.60  ...   
1             1256.17         1031.48            7.05            9.00  ...   
2             1340.13         1105.88            3.91            5.69  ...   
3             1339.70         1107.26            3.91            5.69  ...   
4             1353.04         1117.80            5.48            7.97  ...   
...               ...             ...             ...             ...  ...   
41209         1364.40         1129.17            3.91            5.72  ...   
41210         1494.75         1260.88            9.35           13.66  ...   
41211         1370.65         1130.97            5.48            8.00  ...   
41212         1489.11         1256.25            9.35           13.66  ...   
41213         1378.58         1140.70            5.48            8.00  ...   

       SensorMeasure13  SensorMeasure14  SensorMeasure15  SensorMeasure16  \
0              2387.78          8048.98           9.2229             0.02   
1              2028.09          7863.46          10.8632             0.02   
2              2387.95          8071.13           9.3960             0.02   
3              2387.90          8078.89           9.3594             0.02   
4              2387.87          8057.83           9.3030             0.02   
...                ...              ...              ...              ...   
41209          2388.50          8112.61           9.4427             0.02   
41210          2388.33          8086.83           9.2772             0.02   
41211          2388.64          8100.84           9.3982             0.02   
41212          2388.37          8085.24           9.2727             0.03   
41213          2388.59          8098.17           9.3964             0.02   

       SensorMeasure17  SensorMeasure18  SensorMeasure19  SensorMeasure20  \
0                  362             2324           100.00            24.31   
1                  306             1915            84.93            14.36   
2                  328             2212           100.00            10.39   
3                  328             2212           100.00            10.56   
4                  333             2223           100.00            14.85   
...                ...              ...              ...              ...   
41209              331             2212           100.00            10.53   
41210              366             2324           100.00            24.33   
41211              336             2223           100.00            14.69   
41212              366             2324           100.00            24.44   
41213              335             2223           100.00            14.72   

       SensorMeasure21        LR  
0              14.7007  0.003968  
1               8.5748  0.007937  
2               6.4365  0.011905  
3               6.2367  0.015873  
4               8.9326  0.019841  
...                ...       ...  
41209           6.2620  0.902280  
41210          14.6486  0.905537  
41211           8.8389  0.90

In [93]:
labels=[]
for i in range (0,len(raw_data)):
    if np.array(raw_data["LR"])[i] <= 0.6:
        labels.append(0)
    elif np.array(raw_data["LR"])[i] <= 0.8 :
        labels.append(1)
    else :
        labels.append(2)

raw_data["labels"]=labels
raw_data  = raw_data.drop(columns=['LR'])
raw_data

Cycle   OpSet1  OpSet2  OpSet3  SensorMeasure1  SensorMeasure2  \
0          1  20.0072  0.7000   100.0          491.19          606.67   
1          2  24.9984  0.6200    60.0          462.54          536.22   
2          3  42.0000  0.8420   100.0          445.00          549.23   
3          4  42.0035  0.8402   100.0          445.00          549.19   
4          5  35.0079  0.8400   100.0          449.44          555.10   
...      ...      ...     ...     ...             ...             ...   
41209    277  41.9991  0.8401   100.0          445.00          550.30   
41210    278  20.0026  0.7005   100.0          491.19          608.00   
41211    279  34.9988  0.8413   100.0          449.44          555.92   
41212    280  20.0027  0.7000   100.0          491.19          608.19   
41213    281  35.0075  0.8402   100.0          449.44          556.40   

       SensorMeasure3  SensorMeasure4  SensorMeasure5  SensorMeasure6  ...  \
0             1481.04         1227.81            9.35           13.60  ...   
1             1256.17         1031.48            7.05            9.00  ...   
2             1340.13         1105.88            3.91            5.69  ...   
3             1339.70         1107.26            3.91            5.69  ...   
4             1353.04         1117.80            5.48            7.97  ...   
...               ...             ...             ...             ...  ...   
41209         1364.40         1129.17            3.91            5.72  ...   
41210         1494.75         1260.88            9.35           13.66  ...   
41211         1370.65         1130.97            5.48            8.00  ...   
41212         1489.11         1256.25            9.35           13.66  ...   
41213         1378.58         1140.70            5.48            8.00  ...   

       SensorMeasure13  SensorMeasure14  SensorMeasure15  SensorMeasure16  \
0              2387.78          8048.98           9.2229             0.02   
1              2028.09          7863.46          10.8632             0.02   
2              2387.95          8071.13           9.3960             0.02   
3              2387.90          8078.89           9.3594             0.02   
4              2387.87          8057.83           9.3030             0.02   
...                ...              ...              ...              ...   
41209          2388.50          8112.61           9.4427             0.02   
41210          2388.33          8086.83           9.2772             0.02   
41211          2388.64          8100.84           9.3982             0.02   
41212          2388.37          8085.24           9.2727             0.03   
41213          2388.59          8098.17           9.3964             0.02   

       SensorMeasure17  SensorMeasure18  SensorMeasure19  SensorMeasure20  \
0                  362             2324           100.00            24.31   
1                  306             1915            84.93            14.36   
2                  328             2212           100.00            10.39   
3                  328             2212           100.00            10.56   
4                  333             2223           100.00            14.85   
...                ...              ...              ...              ...   
41209              331             2212           100.00            10.53   
41210              366             2324           100.00            24.33   
41211              336             2223           100.00            14.69   
41212              366             2324           100.00            24.44   
41213              335             2223           100.00            14.72   

       SensorMeasure21  labels  
0              14.7007       0  
1               8.5748       0  
2               6.4365       0  
3               6.2367       0  
4               8.9326       0  
...                ...     ...  
41209           6.2620       2  
41210          14.6486       2  
41211           8.8389       2  
41212        

In [94]:
save_path = '/content/drive/MyDrive/Project_placement/RUL/CMaps/Test_classification_4.csv'
raw_data.to_csv(save_path, index=False)